File: javascript/Hash/algoLadderStarter/arraySubset/arraySubset_Explained.js
// Given two arrays, determine whether one is a subset of the other. It is considered a subset if all the values in one array are contained within the other.

// NOTE: You must accomplish this in O(n) time. This is also known as linear time.

// Input: [1, 2, 3, 4, 5, 6], [6, 3, 2]
// Output: true

// Input: [1, 2, 3, 4, 5, 6], [6, 3, 7]
// Output: false

// Create a function that takes in two input arrays
const arraySubset = function (inputArrayOne, inputArrayTwo) {
  // Determines which input array is longer and saves each input string accordingly
  if (inputArrayOne.length >= inputArrayTwo.length) {
    var longerArray = inputArrayOne
    var shorterArray = inputArrayTwo
  } else {
    var longerArray = inputArrayTwo
    var shorterArray = inputArrayOne
  }
  // Create an empty hash that we can store all the numbers from our longer input array in
  let allNumbersInArray = {}
  // Create our first for loop. This loop will add each element from the inputArrayOne to the hash.
  for (let i = 0; i < longerArray.length; i += 1) {
    if (!allNumbersInArray[longerArray[i]]) {
      allNumbersInArray[longerArray[i]] = true
    }
  }
  // Create our second for loop. This loop will check if each item present in our shorterArray is present in the hash.
  // If it is not we know this is not a subset of the longerArray so return false.
  for (let j = 0; j < shorterArray.length; j += 1) {
    if (!allNumbersInArray[shorterArray[j]]) {
      return false
    }
  }
  // If it passes that for loop with no issues then it is a subset and return true.
  return true
}
// Call our solution with two test inputs, the first one is true while the second is false
console.log(arraySubset([1, 2, 3, 4, 5, 6], [6, 3, 2]))
console.log(arraySubset([1, 2, 3, 4, 5, 6], [6, 3, 7]))

// Although this solution uses two for loops it is still an O(N) solution because they are not nested.

File: ruby/algo_ladder/array_subset/array_subset_explained.rb
# Given two arrays, determine whether one is a subset of the other.
# It is considered a subset if all the values in one array are contained within the other.

# NOTE: You must accomplish this in O(n) time. This is also known as linear time.

# Input: [1, 2, 3, 4, 5, 6], [6, 3, 2]
# Output: true

# Input: [1, 2, 3, 4, 5, 6], [6, 3, 7]
# Output: false

# Create a function that takes in two input arrays
def array_subset(input_array_one, input_array_two)
  # Create a hash to store our first array in
  subset_hash = {}
  # Create an each loop to add each number to our subset_hash
  input_array_one.each do |number|
    subset_hash[number] = true
  end
  # Create an each loop to check if our subset_hash includes our second array.
  # If it does not then return false
  input_array_two.each do |number|
    if !subset_hash[number]
      return false
    end
  end
  # Else return true
  return true
end
# Call our function with two tech arrays
pp array_subset([1, 2, 3, 4, 5, 6], [6, 3, 2])
pp array_subset([1, 2, 3, 4, 5, 6], [6, 3, 7])